pgwire panics when opening a connection from jdbc.postgresql

So I am trying to get clojure (for jepsen, see #4036 ) to talk to cockroach via jdbc and `--insecure`. 
For this I use the standard Java org.postgresql JDBC driver (provided by postgres) version 9.4-1204-jdbc42 and clojure/java.jdbc 0.4.1 (JDK 8)

My JDBC object is instantiated with the following config:
```
(fn conn-spec [_]
                            {:classname   "org.postgresql.Driver"
                             :subprotocol "postgresql"
                             :subname     (str "//n1:15432/system")
                             :user        "psql"
                             :password    "dummy"})
```

Upon starting the client, the (Java) client fails with:
```
WARN  jepsen.cockroach - Lost connection {:classname org.postgresql.Driver, :subprotocol postgresql, :subname //n1:15432/system, :user psql, :password dummy, :connection #object[org.postgresql
.jdbc42.Jdbc42Connection 0x28ba4cca org.postgresql.jdbc42.Jdbc42Connection@28ba4cca]} , reconnecting
```

(then the client gives up)
After this happens, I observe the cockroach server has died with the following trace:

```
I0113 23:16:18.354134 17749 server/node.go:389  connecting to gossip network to verify cluster ID...
I0113 23:16:18.354150 17749 server/node.go:406  node connected via gossip and verified as part of cluster "ddd32b9f-4d8e-42c9-b087-2ce276a97cc7"
I0113 23:16:18.354224 17749 server/node.go:258  Started node with [[ssd]=data] engine(s) and attributes []
I0113 23:16:18.354417 17749 server/server.go:262  starting http server at ip-172-31-63-242:26257
I0113 23:16:18.354528 17749 sql/pgwire/server.go:76  starting postgres server at [::]:15432
W0113 23:16:20.355864 17749 storage/store.go:639  error gossiping system config: range 0: replica <nil> not leader; leader is <nil>
I0113 23:16:28.355472 17749 server/status/runtime.go:137  runtime stats: 60 goroutines, 3.94MiB active, 0.00cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (0x)
W0113 23:16:31.376006 17749 sql/pgwire/v3.go:147  unrecognized configuration parameter "client_encoding"
W0113 23:16:31.376175 17749 sql/pgwire/v3.go:147  unrecognized configuration parameter "DateStyle"
W0113 23:16:31.376311 17749 sql/pgwire/v3.go:147  unrecognized configuration parameter "TimeZone"
W0113 23:16:31.376328 17749 sql/pgwire/v3.go:147  unrecognized configuration parameter "extra_float_digits"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x9c pc=0xb69cfc]

goroutine 127 [running]:
github.com/cockroachdb/cockroach/sql.(*planner).Show(0xc820480500, 0xc8204cd340, 0x0, 0x0, 0x2)
        /go/src/github.com/cockroachdb/cockroach/sql/show.go:47 +0xb9c
github.com/cockroachdb/cockroach/sql.(*planner).makePlan(0xc820480500, 0x7f93d550d978, 0xc8204cd340, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/plan.go:130 +0x63e
github.com/cockroachdb/cockroach/sql.(*Executor).StatementResult(0xc8201d0210, 0xc82034f009, 0x4, 0x7f93d550d978, 0xc8204cd340, 0xc8204e65a0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/executor.go:152 +0x275
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleParse(0xc82005c000, 0xc82005c038, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:321 +0xc31
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc82005c000, 0x0, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:223 +0xb62
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serveConn(0xc8201c3110, 0x7f93d550b228, 0xc8200f41b8, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:199 +0x9b5
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serve.func1(0xc8201c3110, 0x7f93d550b228, 0xc8200f41b8)
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:108 +0x90
created by github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serve
        /go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:113 +0x20a
```

Not sure where to go from here. Is there a way to ask the server to emit the sql queries into its log / stderr?

my failure is in the same place. I'm issuing a `INSERT`, but it fails within `handleParse` on a nil txn.

Last few log lines and trace:
```
W0114 12:01:48.771447 31624 sql/pgwire/v3.go:147  unrecognized configuration parameter "extra_float_digits"
W0114 12:01:48.771475 31624 sql/pgwire/v3.go:147  unrecognized configuration parameter "client_encoding"
W0114 12:01:48.771485 31624 sql/pgwire/v3.go:147  unrecognized configuration parameter "datestyle"
I0114 12:01:48.771556 31624 sql/pgwire/v3.go:182  pgwire: serverMsgReady: 'I'
I0114 12:01:48.771658 31624 sql/pgwire/v3.go:207  pgwire: processing clientMsgParse
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0xa17cbe]

goroutine 228 [running]:
github.com/cockroachdb/cockroach/client.(*Txn).Get(0x0, 0x1498540, 0xc82068d960, 0x0, 0x0, 0x0, 0x0, 0xc800000000)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/client/txn.go:161 +0x1ce
github.com/cockroachdb/cockroach/sql.(*planner).getDescriptor(0xc820214a00, 0x7fad25bf2778, 0xc820667120, 0x7fad25bee5c0, 0xc82068d940, 0x7fad25bf2778)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/descriptor.go:129 +0xb5
github.com/cockroachdb/cockroach/sql.(*planner).getDatabaseDesc(0xc820214a00, 0xc82071700d, 0x6, 0x0, 0x0)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/database.go:48 +0x102
github.com/cockroachdb/cockroach/sql.(*planner).getTableDesc(0xc820214a00, 0xc8205fa510, 0x152e0c0, 0x6)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/table.go:184 +0xe5
github.com/cockroachdb/cockroach/sql.(*planner).getTableLease(0xc820214a00, 0xc8205fa510, 0x5f65646f6e227b5b, 0x73222c313a226469)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/table.go:223 +0x127
github.com/cockroachdb/cockroach/sql.(*planner).Insert(0xc820214a00, 0xc8205fa540, 0xe984f100, 0x0, 0x0, 0x3)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/insert.go:37 +0x9b
github.com/cockroachdb/cockroach/sql.(*planner).makePlan(0xc820214a00, 0x7fad25bf23e8, 0xc8205fa540, 0x0, 0x0, 0x0, 0x77575535582f6632)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/plan.go:106 +0x1151
github.com/cockroachdb/cockroach/sql.(*Executor).StatementResult(0xc8202d0f20, 0xc8205a202e, 0x4, 0x7fad25bf23e8, 0xc8205fa540, 0xc8205fa5a0, 0x0, 0x0, 0x0, 0x0)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/executor.go:152 +0x275
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleParse(0xc8205c64e0, 0xc8205c6518, 0x0, 0x0)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:321 +0xc31
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc8205c64e0, 0x0, 0x0, 0x0)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:223 +0xb62
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serveConn(0xc8202be750, 0x7fad25bf21a0, 0xc8204e8138, 0x0, 0x0)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:199 +0x9b5
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serve.func1(0xc8202be750, 0x7fad25bf21a0, 0xc8204e8138)
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:108 +0x90
created by github.com/cockroachdb/cockroach/sql/pgwire.(*Server).serve
	/home/marc/cockroach/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:113 +0x20a
```

Specifically, it's going into `handleParse` which calls `executor.StatementResult`. However, `StatementResult` does not create a transaction, `executor.ExecuteStatements` does.

--- a/sql/delete.go
+++ b/sql/delete.go
@@ -49,6 +49,10 @@ func (p *planner) Delete(n *parser.Delete) (planNode, *roachpb.Error) {
 		return nil, pErr
 	}
 
+	if p.prepareOnly {
+		return nil, nil
+	}
+
 	// Construct a map from column ID to the index the value appears at within a
 	// row.
 	colIDtoRowIndex, pErr := makeColIDtoRowIndex(rows, tableDesc)

--- a/sql/executor.go
+++ b/sql/executor.go
@@ -184,8 +184,14 @@ func (e *Executor) getSystemConfig() config.SystemConfig {
 	return cfg
 }
 
-// StatementResult returns the result types of the given statement(s).
-func (e *Executor) StatementResult(user string, stmt parser.Statement, args parser.MapArgs) ([]ResultColumn, *roachpb.Error) {
+// Prepare returns the result types of the given statement. Args may be a
+// partially populated val args map. Prepare will populate the missing val
+// args. The column result types are returned (or nil if there are no results).
+func (e *Executor) Prepare(user string, query string, args parser.MapArgs) ([]ResultColumn, *roachpb.Error) {
+	stmt, err := parser.ParseOneTraditional(query)
+	if err != nil {
+		return nil, roachpb.NewError(err)
+	}
 	planMaker := plannerPool.Get().(*planner)
 	defer plannerPool.Put(planMaker)
 
@@ -203,10 +209,16 @@ func (e *Executor) StatementResult(user string, stmt parser.Statement, args pars
 		systemConfig: e.getSystemConfig(),
 	}
 
-	planMaker.evalCtx.StmtTimestamp = parser.DTimestamp{Time: time.Now()}
-	plan, err := planMaker.makePlan(stmt, false)
-	if err != nil {
-		return nil, err
+	timestamp := time.Now()
+	txn := client.NewTxn(e.db)
+	planMaker.setTxn(txn, timestamp)
+	planMaker.evalCtx.StmtTimestamp = parser.DTimestamp{Time: timestamp}
+	plan, pErr := planMaker.prepare(stmt)
+	if pErr != nil {
+		return nil, pErr
+	}
+	if plan == nil {
+		return nil, nil
 	}
 	cols := plan.Columns()
 	for _, c := range cols {

--- a/sql/insert.go
+++ b/sql/insert.go
@@ -161,9 +161,13 @@ func (p *planner) Insert(n *parser.Insert, autoCommit bool) (planNode, *roachpb.
 		for i, val := range rowVals {
 			// Make sure the value can be written to the column before proceeding.
 			var mpErr *roachpb.Error
-			if marshalled[i], mpErr = marshalColumnValue(cols[i], val); mpErr != nil {
+			if marshalled[i], mpErr = marshalColumnValue(cols[i], val, p.evalCtx.Args); mpErr != nil {
 				return nil, mpErr
 			}
+		}
+
+		if p.prepareOnly {
+			continue
 		}
 
 		primaryIndexKey, _, epErr := encodeIndexKey(
@@ -231,6 +235,10 @@ func (p *planner) Insert(n *parser.Insert, autoCommit bool) (planNode, *roachpb.
 	}
 	if pErr := rows.PErr(); pErr != nil {
 		return nil, pErr
+	}
+
+	if p.prepareOnly {
+		return nil, nil
 	}
 
 	if isSystemConfigID(tableDesc.GetID()) {

--- a/sql/parser/builtins.go
+++ b/sql/parser/builtins.go
@@ -1421,7 +1421,7 @@ func typeTuple(params MapArgs, args DTuple) (Datum, error) {
 	}
 	if hasValArgs {
 		for _, arg := range args {
-			_, err := params.setInferredType(arg, datum)
+			_, err := params.SetInferredType(arg, datum)
 			if err != nil {
 				return nil, err
 			}

--- a/sql/parser/eval.go
+++ b/sql/parser/eval.go
@@ -1374,7 +1374,7 @@ func (t Tuple) Eval(ctx EvalContext) (Datum, error) {
 
 // Eval implements the Expr interface.
 func (t ValArg) Eval(_ EvalContext) (Datum, error) {
-	return nil, util.Errorf("unhandled type %T", t)
+	return DValArg{name: t.name}, nil
 }
 
 // Eval implements the Expr interface.

--- a/sql/parser/type_check.go
+++ b/sql/parser/type_check.go
@@ -61,12 +61,12 @@ func (expr *BinaryExpr) TypeCheck(args MapArgs) (Datum, error) {
 	expr.rtype = reflect.TypeOf(dummyRight)
 
 	if expr.ltype == valargType {
-		if _, err := args.setInferredType(dummyLeft, dummyRight); err != nil {
+		if _, err := args.SetInferredType(dummyLeft, dummyRight); err != nil {
 			return nil, err
 		}
 		expr.ltype = expr.rtype
 	} else if expr.rtype == valargType {
-		if _, err := args.setInferredType(dummyRight, dummyLeft); err != nil {
+		if _, err := args.SetInferredType(dummyRight, dummyLeft); err != nil {
 			return nil, err
 		}
 		expr.rtype = expr.ltype
@@ -133,7 +133,7 @@ func (expr *CastExpr) TypeCheck(args MapArgs) (Datum, error) {
 		return nil, err
 	}
 
-	if set, err := args.setInferredType(dummyExpr, DummyString); err != nil {
+	if set, err := args.SetInferredType(dummyExpr, DummyString); err != nil {
 		return nil, err
 	} else if set != nil {
 		dummyExpr = DummyString
@@ -555,7 +555,7 @@ func typeCheckBooleanExprs(args MapArgs, op string, exprs ...Expr) (Datum, error
 		if dummyExpr == DNull {
 			continue
 		}
-		if set, err := args.setInferredType(dummyExpr, DummyBool); err != nil {
+		if set, err := args.SetInferredType(dummyExpr, DummyBool); err != nil {
 			return nil, err
 		} else if set != nil {
 			continue
@@ -568,11 +568,11 @@ func typeCheckBooleanExprs(args MapArgs, op string, exprs ...Expr) (Datum, error
 }
 
 func typeCheckComparisonOp(args MapArgs, op ComparisonOp, dummyLeft, dummyRight Datum) (Datum, cmpOp, error) {
-	if set, err := args.setInferredType(dummyLeft, dummyRight); err != nil {
+	if set, err := args.SetInferredType(dummyLeft, dummyRight); err != nil {
 		return nil, cmpOp{}, err
 	} else if set != nil {
 		dummyLeft = set
-	} else if set, err := args.setInferredType(dummyRight, dummyLeft); err != nil {
+	} else if set, err := args.SetInferredType(dummyRight, dummyLeft); err != nil {
 		return nil, cmpOp{}, err
 	} else if set != nil {
 		dummyRight = set

--- a/sql/parser/walk.go
+++ b/sql/parser/walk.go
@@ -237,11 +237,11 @@ func (m MapArgs) Arg(name string) (Datum, bool) {
 	return d, ok
 }
 
-// setValArg sets the bind var argument d to the type typ in m. If m is nil
-// or d is not a DValArg, nil is returned. If the bind var argument is set,
+// SetInferredType sets the bind var argument d to the type typ in m. If m is
+// nil or d is not a DValArg, nil is returned. If the bind var argument is set,
 // typ is returned. An error is returned if typ cannot be set because a
 // different type is already present.
-func (m MapArgs) setInferredType(d, typ Datum) (set Datum, err error) {
+func (m MapArgs) SetInferredType(d, typ Datum) (set Datum, err error) {
 	if m == nil {
 		return nil, nil
 	}
@@ -366,25 +366,4 @@ func containsSubquery(expr Expr) bool {
 	v := containsSubqueryVisitor{containsSubquery: false}
 	_ = WalkExpr(&v, expr)
 	return v.containsSubquery
-}
-
-type checkVisitor struct {
-	args MapArgs
-	err  error
-}
-
-func (v *checkVisitor) Visit(expr Expr, pre bool) (Visitor, Expr) {
-	if !pre || v.err != nil {
-		return nil, expr
-	}
-	_, v.err = expr.TypeCheck(v.args)
-	// Return nil to stop recursion since TypeCheck above recurses.
-	return nil, expr
-}
-
-// InferArgs populates args with the inferred types of stmt's placeholders.
-func InferArgs(stmt Statement, args MapArgs) error {
-	v := checkVisitor{args: args}
-	WalkStmt(&v, stmt)
-	return v.err
 }

--- a/sql/pgwire/v3.go
+++ b/sql/pgwire/v3.go
@@ -302,10 +302,6 @@ func (c *v3Conn) handleParse(buf *readBuffer) error {
 		}
 		inTypeHints[i] = oid.Oid(typ)
 	}
-	stmt, err := parser.ParseOneTraditional(query)
-	if err != nil {
-		return c.sendError(err.Error())
-	}
 	args := make(parser.MapArgs)
 	for i, t := range inTypeHints {
 		if t == 0 {
@@ -317,13 +313,15 @@ func (c *v3Conn) handleParse(buf *readBuffer) error {
 		}
 		args[fmt.Sprint(i+1)] = v
 	}
-	if err := parser.InferArgs(stmt, args); err != nil {
-		return c.sendError(err.Error())
+	cols, pErr := c.executor.Prepare(c.opts.user, query, args)
+	if pErr != nil {
+		return c.sendError(pErr.String())
 	}
 	pq := preparedStatement{
 		query:       query,
 		inTypes:     make([]oid.Oid, len(args)),
 		portalNames: make(map[string]struct{}),
+		columns:     cols,
 	}
 	copy(pq.inTypes, inTypeHints)
 	for k, v := range args {
@@ -342,11 +340,6 @@ func (c *v3Conn) handleParse(buf *readBuffer) error {
 		}
 		pq.inTypes[i-1] = id
 	}
-	cols, pErr := c.executor.StatementResult(c.opts.user, stmt, args)
-	if pErr != nil {
-		return c.sendError(pErr.String())
-	}
-	pq.columns = cols
 	c.preparedStatements[name] = pq
 	c.writeBuf.initMsg(serverMsgParseComplete)
 	return c.writeBuf.finishMsg(c.wr)

--- a/sql/pgwire_test.go
+++ b/sql/pgwire_test.go
@@ -150,9 +150,8 @@ func TestPGWire(t *testing.T) {
 }
 
 type preparedTest struct {
-	params []interface{}
-	error  string
-	result []interface{}
+	params, results []interface{}
+	error           string
 }
 
 func (p preparedTest) Params(v ...interface{}) preparedTest {
@@ -166,11 +165,11 @@ func (p preparedTest) Error(err string) preparedTest {
 }
 
 func (p preparedTest) Results(v ...interface{}) preparedTest {
-	p.result = v
+	p.results = v
 	return p
 }
 
-func TestPGPrepared(t *testing.T) {
+func TestPGPreparedQuery(t *testing.T) {
 	defer leaktest.AfterTest(t)
 	var base preparedTest
 	queryTests := map[string][]preparedTest{
@@ -219,6 +218,10 @@ func TestPGPrepared(t *testing.T) {
 			base.Params(3, "4").Results(6, 7),
 			base.Params(0, "a").Error(`pq: param $2: strconv.ParseInt: parsing "a": invalid syntax`),
 		},
+		// Check for QualifiedName resolution.
+		"SELECT COUNT(*)": {
+			base.Results(1),
+		},
 		// TODO(mjibson): test date/time types
 	}
 
@@ -241,52 +244,46 @@ func TestPGPrepared(t *testing.T) {
 			continue
 		}
 
-		type result struct {
-			rows *sql.Rows
-			err  error
-		}
-
 		for _, test := range tests {
-			var results []result
-			{
-				rows, err := db.Query(query, test.params...)
-				results = append(results, result{rows: rows, err: err})
-			}
-			{
-				rows, err := stmt.Query(test.params...)
-				results = append(results, result{rows: rows, err: err})
-			}
-			for _, res := range results {
-				rows, err := res.rows, res.err
+			for _, queryFunc := range []func(...interface{}) (*sql.Rows, error){
+				func(args ...interface{}) (*sql.Rows, error) {
+					return db.Query(query, args...)
+				},
+				stmt.Query,
+			} {
+				rows, err := queryFunc(test.params...)
 				if err != nil {
 					if test.error == "" {
 						t.Errorf("%s: %#v: unexpected error: %s", query, test.params, err)
-					}
-					if test.error != err.Error() {
+					} else if test.error != err.Error() {
 						t.Errorf("%s: %#v: expected error: %s, got %s", query, test.params, test.error, err)
 					}
 					continue
 				}
 				if test.error != "" && err == nil {
 					t.Errorf("expected error: %s: %#v", query, test.params)
+					continue
 				}
-				dst := make([]interface{}, len(test.result))
-				for i, d := range test.result {
+				dst := make([]interface{}, len(test.results))
+				for i, d := range test.results {
 					dst[i] = reflect.New(reflect.TypeOf(d)).Interface()
 				}
 				if !rows.Next() {
 					t.Errorf("expected row: %s: %#v", query, test.params)
+					continue
 				}
 				if err := rows.Scan(dst...); err != nil {
 					t.Error(err)
+					continue
 				}
 				rows.Close()
 				for i, d := range dst {
 					v := reflect.Indirect(reflect.ValueOf(d)).Interface()
 					dst[i] = v
 				}
-				if !reflect.DeepEqual(dst, test.result) {
-					t.Errorf("%s: %#v: expected %v, got %v", query, test.params, test.result, dst)
+				if !reflect.DeepEqual(dst, test.results) {
+					t.Errorf("%s: %#v: expected %v, got %v", query, test.params, test.results, dst)
+					continue
 				}
 			}
 		}
@@ -312,6 +309,94 @@ func TestPGPrepared(t *testing.T) {
 		}
 		if err.Error() != reason {
 			t.Errorf("unexpected error: %s: %s", query, err)
+		}
+	}
+}
+
+func TestPGPreparedExec(t *testing.T) {
+	defer leaktest.AfterTest(t)
+
+	var base preparedTest
+	queryTests := []struct {
+		query string
+		tests []preparedTest
+	}{
+		{
+			"INSERT INTO d.t VALUES ($1, $2, $3)",
+			[]preparedTest{
+				base.Params(1, "one", 2),
+				base.Params("two", 2, 2).Error(`pq: param $1: strconv.ParseInt: parsing "two": invalid syntax`),
+			},
+		},
+		{
+			"UPDATE d.t SET s = $1, i = i + $2, d = 1 + $3 WHERE i = $4",
+			[]preparedTest{
+				base.Params(4, 3, 2, 1),
+			},
+		},
+		{
+			"DELETE FROM d.t WHERE s = $1 and i = $2 and d = 2 + $3",
+			[]preparedTest{
+				base.Params(1, 2, 3),
+			},
+		},
+	}
+
+	s := server.StartTestServer(t)
+	defer s.Stop()
+
+	pgUrl, cleanupFn := sqlutils.PGUrl(t, s, security.RootUser, os.TempDir(), "TestPGPrepared")
+	defer cleanupFn()
+
+	db, err := sql.Open("postgres", pgUrl.String())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+
+	setup := `CREATE DATABASE d; CREATE TABLE d.t (i INT, s STRING, d INT);`
+	if _, err := db.Exec(setup); err != nil {
+		t.Fatal(err)
+	}
+
+	for _, queryTest := range queryTests {
+		stmt, err := db.Prepare(queryTest.query)
+		if err != nil {
+			t.Fatalf("prepare error: %s: %s", queryTest.query, err)
+		}
+		for _, test := range queryTest.tests {
+			_, err := stmt.Exec(test.params...)
+			if err != nil {
+				if test.error == "" {
+					t.Fatalf("%s: %#v: unexpected error: %s", queryTest.query, test.params, err)
+				} else if test.error != err.Error() {
+					t.Fatalf("%s: %#v: expected error: %s, got %s", queryTest.query, test.params, test.error, err)
+				}
+			}
+			if test.error != "" && err == nil {
+				t.Fatalf("expected error: %s: %#v", queryTest.query, test.params)
+			}
+		}
+		if err := stmt.Close(); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	testFailures := map[string]string{
+		"CREATE TABLE $1 (id INT)":  "pq: syntax error at or near \"1\"\nCREATE TABLE $1 (id INT)\n             ^\n",
+		"DROP TABLE t":              "pq: prepare statement not supported: *parser.DropTable",
+		"UPDATE d.t SET s = i + $1": "pq: value type int doesn't match type STRING of column \"s\"",
+	}
+
+	for query, reason := range testFailures {
+		stmt, err := db.Prepare(query)
+		if err == nil {
+			t.Errorf("expected error: %s", query)
+			stmt.Close()
+			continue
+		}
+		if err.Error() != reason {
+			t.Errorf("unexpected error: %s: %q", query, err)
 		}
 	}
 }

--- a/sql/plan.go
+++ b/sql/plan.go
@@ -43,6 +43,10 @@ type planner struct {
 	// on these schema changers to roll out the new schema.
 	schemaChangers []SchemaChanger
 
+	// TODO(mjibson): remove prepareOnly in favor of a 2-step prepare-exec solution
+	// that is also able to save the plan to skip work during the exec step.
+	prepareOnly bool
+
 	testingVerifyMetadata func(config.SystemConfig) error
 
 	parser             parser.Parser
@@ -149,6 +153,26 @@ func (p *planner) makePlan(stmt parser.Statement, autoCommit bool) (planNode, *r
 		return p.Values(n)
 	default:
 		return nil, roachpb.NewErrorf("unknown statement type: %T", stmt)
+	}
+}
+
+func (p *planner) prepare(stmt parser.Statement) (planNode, *roachpb.Error) {
+	p.prepareOnly = true
+	switch n := stmt.(type) {
+	case *parser.Delete:
+		return p.Delete(n)
+	case *parser.Insert:
+		return p.Insert(n, false)
+	case *parser.Select:
+		return p.Select(n)
+	case *parser.Update:
+		return p.Update(n)
+	default:
+		// TODO(mjibson): use stmt.PGTag() instead of %T
+		return nil, roachpb.NewUErrorf("prepare statement not supported: %T", stmt)
+
+		// TODO(mjibson): add support for parser.Values.
+		// Broken because it conflicts with INSERT's use of VALUES.
 	}
 }
 

--- a/sql/table.go
+++ b/sql/table.go
@@ -667,7 +667,7 @@ func encodeSecondaryIndexes(tableID ID, indexes []IndexDescriptor,
 // marshalColumnValue returns a Go primitive value equivalent of val, of the
 // type expected by col. If val's type is incompatible with col, or if
 // col's type is not yet implemented, an error is returned.
-func marshalColumnValue(col ColumnDescriptor, val parser.Datum) (interface{}, *roachpb.Error) {
+func marshalColumnValue(col ColumnDescriptor, val parser.Datum, args parser.MapArgs) (interface{}, *roachpb.Error) {
 	if val == parser.DNull {
 		return nil, nil
 	}
@@ -677,37 +677,85 @@ func marshalColumnValue(col ColumnDescriptor, val parser.Datum) (interface{}, *r
 		if v, ok := val.(parser.DBool); ok {
 			return bool(v), nil
 		}
+		if set, err := args.SetInferredType(val, parser.DummyBool); err != nil {
+			return nil, roachpb.NewError(err)
+		} else if set != nil {
+			return nil, nil
+		}
 	case ColumnType_INT:
 		if v, ok := val.(parser.DInt); ok {
 			return int64(v), nil
+		}
+		if set, err := args.SetInferredType(val, parser.DummyInt); err != nil {
+			return nil, roachpb.NewError(err)
+		} else if set != nil {
+			return nil, nil
 		}
 	case ColumnType_FLOAT:
 		if v, ok := val.(parser.DFloat); ok {
 			return float64(v), nil
 		}
+		if set, err := args.SetInferredType(val, parser.DummyFloat); err != nil {
+			return nil, roachpb.NewError(err)
+		} else if set != nil {
+			return nil, nil
+		}
 	case ColumnType_DECIMAL:
 		if v, ok := val.(parser.DDecimal); ok {
 			return v.Decimal, nil
+		}
+		if set, err := args.SetInferredType(val, parser.DummyDecimal); err != nil {
+			return nil, roachpb.NewError(err)
+		} else if set != nil {
+			return nil, nil
 		}
 	case ColumnType_STRING:
 		if v, ok := val.(parser.DString); ok {
 			return string(v), nil
 		}
+		if set, err := args.SetInferredType(val, parser.DummyString); err != nil {
+			return nil, roachpb.NewError(err)
+		} else if set != nil {
+			return nil, nil
+		}
 	case ColumnType_BYTES:
 		if v, ok := val.(parser.DBytes); ok {
 			return string(v), nil
+		}
+		if v, ok := val.(parser.DString); ok {
+			return string(v), nil
+		}
+		if set, err := args.SetInferredType(val, parser.DummyBytes); err != nil {
+			return nil, roachpb.NewError(err)
+		} else if set != nil {
+			return nil, nil
 		}
 	case ColumnType_DATE:
 		if v, ok := val.(parser.DDate); ok {
 			return int64(v), nil
 		}
+		if set, err := args.SetInferredType(val, parser.DummyDate); err != nil {
+			return nil, roachpb.NewError(err)
+		} else if set != nil {
+			return nil, nil
+		}
 	case ColumnType_TIMESTAMP:
 		if v, ok := val.(parser.DTimestamp); ok {
 			return v.Time, nil
 		}
+		if set, err := args.SetInferredType(val, parser.DummyTimestamp); err != nil {
+			return nil, roachpb.NewError(err)
+		} else if set != nil {
+			return nil, nil
+		}
 	case ColumnType_INTERVAL:
 		if v, ok := val.(parser.DInterval); ok {
 			return v.Duration, nil
+		}
+		if set, err := args.SetInferredType(val, parser.DummyInterval); err != nil {
+			return nil, roachpb.NewError(err)
+		} else if set != nil {
+			return nil, nil
 		}
 	default:
 		return nil, roachpb.NewErrorf("unsupported column type: %s", col.Type.Kind)

--- a/sql/update.go
+++ b/sql/update.go
@@ -132,6 +132,55 @@ func (p *planner) Update(n *parser.Update) (planNode, *roachpb.Error) {
 		return nil, pErr
 	}
 
+	// ValArgs have their types populated in the above Select if they are part
+	// of an expression ("SET a = 2 + $1") in the type check step where those
+	// types are inferred. For the simpler case ("SET a = $1"), populate them
+	// using marshalColumnValue. This step also verifies that the expression
+	// types match the column types.
+	if p.prepareOnly {
+		i := 0
+		f := func(expr parser.Expr) *roachpb.Error {
+			idx := i
+			i++
+			// DefaultVal doesn't implement TypeCheck
+			if _, ok := expr.(parser.DefaultVal); ok {
+				return nil
+			}
+			d, err := expr.TypeCheck(nil)
+			if err != nil {
+				return roachpb.NewError(err)
+			}
+			if _, err := marshalColumnValue(cols[idx], d, p.evalCtx.Args); err != nil {
+				return err
+			}
+			return nil
+		}
+		for _, expr := range n.Exprs {
+			if expr.Tuple {
+				switch t := expr.Expr.(type) {
+				case parser.Tuple:
+					for _, e := range t {
+						if err := f(e); err != nil {
+							return nil, err
+						}
+					}
+				case parser.DTuple:
+					for _, e := range t {
+						if err := f(e); err != nil {
+							return nil, err
+						}
+					}
+				}
+			} else {
+				if err := f(expr.Expr); err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		return nil, nil
+	}
+
 	// Construct a map from column ID to the index the value appears at within a
 	// row.
 	colIDtoRowIndex := map[ColumnID]int{}
@@ -216,7 +265,7 @@ func (p *planner) Update(n *parser.Update) (planNode, *roachpb.Error) {
 		// cannot be used as index values.
 		for i, val := range newVals {
 			var mpErr *roachpb.Error
-			if marshalled[i], mpErr = marshalColumnValue(cols[i], val); mpErr != nil {
+			if marshalled[i], mpErr = marshalColumnValue(cols[i], val, p.evalCtx.Args); mpErr != nil {
 				return nil, mpErr
 			}
 		}

